xla/service/gpu/cudnn_norm_rewriter_test.cc

```diff
@@ -620,6 +620,472 @@ TEST_F(CudnnNormRewriterTest, LayerNormTrain4D12) {
+      a = f32[] parameter(0)
+      b = f32[] parameter(1)
+      ROOT c = f32[] add(a,b)
+    }
+
+    ENTRY test {
+        input = f32[2,4] parameter(0)
+        input_square = f32[2,4] multiply(input, input)
+        c0 = f32[] constant(0)
+        input_square_sum = f32[2] reduce(input_square, c0), dimensions={1}, to_apply=apply
+        reduce = f32[2] reduce(input, c0), dimensions={1}, to_apply=apply
+        r_nelems = f32[] constant(0.25)
+        r_nelems_bcast = f32[2] broadcast(r_nelems), dimensions={}
+        input_square_mean = f32[2] multiply(input_square_sum,r_nelems_bcast)
+        input_mean = f32[2] multiply(reduce, r_nelems_bcast)
+        input_mean_square = f32[2] multiply(input_mean,input_mean)
+        variance = f32[2] subtract(input_square_mean,input_mean_square)
+        epsilon = f32[] constant(0.001)
+        epsilon_bcast = f32[2] broadcast(epsilon), dimensions={}
+        variance_plus_epsilon = f32[2] add(variance, epsilon_bcast)
+        norm_factor = f32[2] rsqrt(variance_plus_epsilon)
+        norm_factor_bcast = f32[2,4] broadcast(norm_factor), dimensions={0}
+        input_mean_bcast = f32[2,4] broadcast(input_mean), dimensions={0}
+        input_center = f32[2,4] subtract(input, input_mean_bcast)
+        norm = f32[2,4] multiply(input_center, norm_factor_bcast)
+        scale = f32[4] parameter(1)
+        scale_bcast = f32[2,4] broadcast(scale), dimensions={1}
+        norm_scale = f32[2,4] multiply(norm, scale_bcast)
+        bias = f32[4] parameter(2)
+        bias_bcast = f32[2,4] broadcast(bias), dimensions={1}
+        norm_scale_bias = f32[2,4] add(norm_scale, bias_bcast)
+        doutput = f32[2,4] parameter(3)
+        dbias = f32[4] reduce(doutput, c0), dimensions={0}, to_apply=apply
+        norm_doutput = f32[2,4] multiply(norm, doutput)
+        dscale = f32[4] reduce(norm_doutput, c0), dimensions={0}, to_apply=apply
+        scale_doutput = f32[2,4] multiply(scale_bcast, doutput)
+        input_center_scale_doutput = f32[2,4] multiply(input_center, scale_doutput)
+        f0 = f32[2] reduce(input_center_scale_doutput, c0), dimensions={1}, to_apply=apply
+        norm_factor_cube = f32[2] divide(norm_factor, variance_plus_epsilon)
+        c1 = f32[] constant(-0.5)
+        c1_bcast = f32[2] broadcast(c1), dimensions={}
+        dnorm_factor = f32[2] multiply(norm_factor_cube, c1_bcast)
+        f0_dnorm_factor = f32[2] multiply(f0, dnorm_factor)
+        c2 = f32[] constant(0.5)
+        c2_bcast = f32[2] broadcast(c2), dimensions={}
+        f0_dnorm_factor_scaled = f32[2] multiply(f0_dnorm_factor, c2_bcast)
+        f0_dnorm_factor_scaled_bcast = f32[2,4] broadcast(f0_dnorm_factor_scaled), dimensions={0}
+        f1 = f32[2,4] multiply(input_center, f0_dnorm_factor_scaled_bcast)
+        minus_f1 = f32[2,4] negate(f1)
+        minus_f1_sum = f32[2] reduce(minus_f1, c0), dimensions={1}, to_apply=apply
+        norm_factor_scale_doutput = f32[2,4] multiply(norm_factor_bcast, scale_doutput)
```

Comment:
Call this f2. And similarly for the other variable names that have the string "norm_factor_scale_doutput". And similarly for other tests.